extensions/machine-learning-services/src/packageManagement/packageManager.ts

```diff
@@ -0,0 +1,166 @@
+				defaultProviderId: SqlPythonPackageManageProvider.ProviderId
+			});
+		} else {
+			this._apiWrapper.showInfoMessage(constants.managePackageCommandError);
+		}
+	}
+
+	/**
+	 * Installs dependencies for the extension
+	 */
+	public async installDependencies(): Promise<void> {
+		let installCompletion = new Deferred<void>();
+		let msgTaskName = constants.installDependenciesMsgTaskName;
+		this._apiWrapper.startBackgroundOperation({
+			displayName: msgTaskName,
+			description: msgTaskName,
+			isCancelable: false,
+			operation: async op => {
+				try {
+					if (!(await utils.exists(this._pythonExecutable))) {
+
+						// Install python
+						//
+						await utils.createFolder(this._pythonInstallationLocation);
+						await this.jupyterInstallation.installPythonPackage(op, false, this._pythonInstallationLocation, this._outputChannel);
+					}
+
+					// Install required packages
+					//
+					await this.installRequiredPythonPackages();
+					op.updateStatus(azdata.TaskStatus.Succeeded);
+					installCompletion.resolve();
+				} catch (error) {
+					let errorMsg = `${constants.installDependenciesError}' ${error}'`;
+					op.updateStatus(azdata.TaskStatus.Failed, errorMsg);
+					installCompletion.reject(errorMsg);
+				}
+			}
+		});
+
+		return installCompletion.promise;
+	}
+
+	/**
+	 * Installs required python packages
+	 */
+	private async installRequiredPythonPackages(): Promise<void> {
+		let installedPackages = await this.getInstalledPipPackages();
+		let fileContent = '';
+		requiredPackages.forEach(packageDetails => {
+			let hasVersion = ('version' in packageDetails);
+			if (!installedPackages.find(x => x.name === packageDetails['name'] && (!hasVersion || packageDetails['version'] === x.version))) {
+				let packageNameDetail = hasVersion ? `${packageDetails.name}==${packageDetails.version}` : `${packageDetails.name}`;
+				fileContent = `${fileContent}${packageNameDetail}\n`;
+			}
+		});
+
+		if (fileContent) {
+			this._outputChannel.appendLine(constants.installDependenciesPackages);
+			let result = await utils.execCommandOnTempFile<string>(fileContent, async (tempFilePath) => {
+				return await this.installPackages(tempFilePath);
+			});
+			this._outputChannel.appendLine(result);
+		} else {
+			this._outputChannel.appendLine(constants.installDependenciesPackagesAlreadyInstalled);
+		}
+	}
+
+	private async getInstalledPipPackages(): Promise<nbExtensionApis.IPackageDetails[]> {
+		try {
+			let cmd = `"${this._pythonExecutable}" -m pip list --format=json`;
+			let packagesInfo = await this.jupyterInstallation.executeBufferedCommand(cmd);
+			let packagesResult: nbExtensionApis.IPackageDetails[] = [];
+			if (packagesInfo) {
+				packagesResult = <nbExtensionApis.IPackageDetails[]>JSON.parse(packagesInfo);
+			}
+			return packagesResult;
+		}
+		catch (err) {
+			this._outputChannel.appendLine(`${constants.installDependenciesGetPackagesError} '${err}'`);
```

Comment:
>appendLine [](start = 23, length = 10)

Just have this be a function that takes in the error and uses a format string to insert it, makes it easier to reuse and see how the final message is going to look.

Example : [https://github.com/microsoft/azuredatastudio/blob/master/extensions/big-data-cluster/src/bigDataCluster/localizedConstants.ts#L22](https://github.com/microsoft/azuredatastudio/blob/master/extensions/big-data-cluster/src/bigDataCluster/localizedConstants.ts#L22)